core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -181,26 +178,15 @@ public static Options defaultOptions() {
   private final int mMaxCapacity;
   private final int mMinCapacity;
 
-  // Tracks the resources that are available ordered by lastAccessTime (the first one is
+  // Tracks the resources that are available ordered by lastAccessTime (the head is
   // the most recently used resource).
   @GuardedBy("mLock")
-  private TreeSet<ResourceInternal<T>> mResourceAvailable =
-      new TreeSet<>(new Comparator<ResourceInternal<T>>() {
-        @Override
-        public int compare(ResourceInternal<T> c1, ResourceInternal<T> c2) {
-          if (c1 == c2) {
-            return 0;
-          }
-          if (c1.mLastAccessTimeMs == c2.mLastAccessTimeMs) {
-            return c1.mIdentity - c2.mIdentity;
-          }
-          return (int) (c2.mLastAccessTimeMs - c1.mLastAccessTimeMs);
-        }
-      });
+  private Deque<ResourceInternal<T>> mResourceAvailable;
 
   // Tracks all the resources that are not closed.
-  @GuardedBy("mLock")
-  private HashMap<T, ResourceInternal<T>> mResources = new HashMap<>(32);
+  // The size of the map is guarded by mLock. The mLock should be acquired if you add/delete/iterate
```

Comment:
ok, I think I get it. I think the comment is misleading. The `mLock` is required to protect the other deque, since that is not a thread-safe data structure. However, it looks like we are also overloading that lock to protect any multi-data-structure operations between `mResources` and `mResourceAvailable`.

Is that correct? If so, that means not every modification to `mResources` requires the lock. Only the accesses that needs transactional semantics between `mResources` and `mResourceAvailable` need the lock. Is this true?